libsolidity/analysis/DeclarationTypeChecker.cpp

```diff
@@ -393,13 +394,37 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
 
 }
 
-void DeclarationTypeChecker::endVisit(UsingForDirective const& _usingFor)
+bool DeclarationTypeChecker::visit(UsingForDirective const& _usingFor)
 {
 	ContractDefinition const* library = dynamic_cast<ContractDefinition const*>(
 		_usingFor.libraryName().annotation().referencedDeclaration
 	);
+
 	if (!library || !library->isLibrary())
 		m_errorReporter.fatalTypeError(4357_error, _usingFor.libraryName().location(), "Library name expected.");
+
+	_usingFor.libraryName().annotation().type = TypeProvider::contract(*library);
+	if (_usingFor.typeName())
+		_usingFor.typeName()->accept(*this);
```

Comment:
How will this not end up in `DeclarationTypeChecker::endVisit(UserDefinedTypeName const& _typeName)` which gives an error for libraries?

Reply:
It doesn't visit ``_usingFor.libraryName()``, but ``_usingFor.typeName()``.
I.e. that's the ``uint256`` not the ``L`` in ``using L for uint256;``.

Reply:
Ah ok